app/src/heartbeat/heartbeat_controller_desktop.cc

```diff
@@ -79,5 +97,171 @@ void HeartbeatController::LogHeartbeat() {
+      // payload and don't attempt to write.
+      if (read_succeeded && stored_heartbeats.heartbeats.size() > 0) {
+        // Find a logged heartbeat from today. Remove it from the stored
+        // heartbeats and use it to construct a single-heartbeat payload.
+        LoggedHeartbeats todays_heartbeats;
+        for (auto& entry : stored_heartbeats.heartbeats) {
+          std::string user_agent = entry.first;
+          std::vector<std::string>& dates = entry.second;
+          auto itr = std::find(dates.begin(), dates.end(), current_date);
+          if (itr != dates.end()) {
+            dates.erase(itr);
+            todays_heartbeats.heartbeats[user_agent].push_back(current_date);
+          }
+        }
+        bool write_succeeded = this->storage_.Write(stored_heartbeats);
+        // Only update last-logged date and return a payload if the write
+        // succeeds.
+        if (write_succeeded) {
+          this->last_fetched_todays_heartbeat_date_ = current_date;
+          *output_str = GetStringPayloadForHeartbeats(todays_heartbeats);
+        }
+      }
+    }
+    // Post the semaphore to signal the main thread to read the result.
+    scheduled_work_semaphore->Post();
+    return;
+  };
+
+  scheduler_.Schedule(get_and_reset_function);
+  // Wait until the scheduled work completes.
+  if (scheduled_work_semaphore->TimedWait(kMaxWaitTimeMs)) {
+    return *output_str;
+  }
+  // Return an empty string if unable to TimedWait times out.
+  return "";
+}
+
+std::string HeartbeatController::GetStringPayloadForHeartbeats(
+    LoggedHeartbeats heartbeats) {
+  Variant heartbeats_variant = Variant::EmptyVector();
+  std::vector<Variant>& heartbeats_vector = heartbeats_variant.vector();
+
+  // heartbeats is a map from user agent to a vector of date strings.
+  for (auto const& entry : heartbeats.heartbeats) {
+    Variant heartbeat_entry = Variant::EmptyMap();
+    std::map<Variant, Variant>& heartbeat_entry_map = heartbeat_entry.map();
+    Variant dates_variant = Variant::EmptyVector();
+    std::vector<Variant>& dates_vector = dates_variant.vector();
+    std::string user_agent = entry.first;
+    for (std::string date : entry.second) {
+      dates_vector.push_back(date);
+    }
+    heartbeat_entry_map[kUserAgentKey] = user_agent;
+    heartbeat_entry_map[kDatesKey] = dates_variant;
+    heartbeats_vector.push_back(heartbeat_entry);
+  }
+  // Construct a final object with 'heartbeats' and 'version'
+  Variant root = Variant::EmptyMap();
+  std::map<Variant, Variant>& root_map = root.map();
+  root_map[kHeartbeatsKey] = heartbeats_variant;
+  root_map[kVersionKey] = "2";
+
+  std::string json_object = util::VariantToJson(root);
+  return CompressAndEncode(json_object);
+}
+
+std::string HeartbeatController::CompressAndEncode(const std::string& input) {
+  ZLib zlib;
+  zlib.SetGzipHeaderMode();
+  uLongf result_size = ZLib::MinCompressbufSize(input.length());
+  std::unique_ptr<char[]> result(new char[result_size]);
```

Comment:
Done.

Also, as a full disclaimer, I basically copied this block of code from 
https://github.com/firebase/firebase-cpp-sdk/blob/main/app/rest/tests/response_binary_test.cc
So I'm not sure if there is any other simpler way to use gzip

Reply:
UniquePtr currently has an assertion indicating that is it not yet supported for array[] types, so I switched it back to unique_ptr for now.
Since I just copied the logic I'm not sure if unique_ptr is even necessary, it seems like a regular pointer might be fine.